src/io/import/importDataSources.ts

```diff
@@ -58,34 +66,89 @@ const unhandledResource: ImportHandler = () => {
 };
 
 export async function importDataSources(dataSources: DataSource[]) {
-  const middleware: Array<ImportHandler> = [
+  const importContext = {
+    fetchFileCache: new Map<string, File>(),
+    dicomDataSources: [] as DataSourceWithFile[],
+  };
+
+  const prioritizes = [
     // updating the file type should be first in the pipeline
     updateFileMimeType,
-    // handleConfig must before restoreStateFile for label props to be applied to deserialized tools
-    handleConfig,
-    restoreStateFile,
+    // before extractArchive as .zip file is part of state file check
+    prioritizeStateFile,
+
+    // source generators
     handleRemoteManifest,
     handleGoogleCloudStorage,
     handleAmazonS3,
     downloadUrl,
     extractArchiveTargetFromCache,
     extractArchive,
+
+    // sinks
+    prioritizeJSON,
+    earliestPriority,
+  ];
+  const prioritizer = new Pipeline<
+    DataSource,
+    PriorityResult | ImportResult,
+    typeof importContext
+  >(prioritizes);
+
+  const priorityOrImportResult = await Promise.all(
+    dataSources.map((r) => prioritizer.execute(r, importContext))
+  );
```

Comment:
I think a simpler approach would be to add a `loadPriority?: number` field to `ImportResult`, then sort on the priority (defaulting to 0 if the key doesn't exist). The priority conversion pipeline + buckets can then be scrapped for a single sort call (or bucket approach). I think it'll be easier to understand and reason about that way.

Reply:
Yeah, what a mess.  But better now.

Ran with your "do it like dicom sources" idea.  `handleConfig` just calls `done({ config ])`.  And the pipeline following code calls `applyConfig` on all resulting `ImportResults` with configs.